### ISSUE TITLE
No line drawn in LineChart when highest DataEntry.x is smaller than xAxis.axisMaximum

### ISSUE THREAD

**Issue by jeroenrietveld** [danielgindi/Charts#4643]
* [x] I've read, understood, and done my best to follow the [*CONTRIBUTING guidelines](https://github.com/jjatie/Charts/blob/master/CONTRIBUTING.md).

## What did you do?

I have a chart where the xAxis maximum is sometimes bigger than the highest xValue in my dataset.
When trying to plot this, no line would show up (LineChart).

This is not an issue when the line is prepended with empty space.

## What did you expect to happen?

I expect the dataset to be draw normally, but only take up the view partially.

## What happened instead?

No line gets drawn.

## Charts Environment

**Charts version/Branch/Commit Number:** 4.0.1
**Xcode version:** 12.5
**Swift version:** 5
**Platform(s) running Charts:**
**macOS version running Xcode:**

## Demo Project
-

## Issue

The line is not drawn, because of an incorrect `XBounds.set`.
```
        open func set(chart: BarLineScatterCandleBubbleChartDataProvider,
                      dataSet: BarLineScatterCandleBubbleChartDataSetProtocol,
                      animator: Animator?)
        {
            let phaseX = Swift.max(0.0, Swift.min(1.0, animator?.phaseX ?? 1.0))
            
            let low = chart.lowestVisibleX
            let high = chart.highestVisibleX
            
            let entryFrom = dataSet.entryForXValue(low, closestToY: .nan, rounding: .down)
            let entryTo = dataSet.entryForXValue(high, closestToY: .nan, rounding: .up)
            
            self.min = entryFrom == nil ? 0 : dataSet.entryIndex(entry: entryFrom!)
            self.max = entryTo == nil ? 0 : dataSet.entryIndex(entry: entryTo!)
            range = Int(Double(self.max - self.min) * phaseX)
        }
```
The `entryTo` does not the latest dataentry, but nil. This results in bounds from min: 0, to max: 0.

`entryForXValue` returns nil because `index == -1`:
```
    open override func entryForXValue(
        _ xValue: Double,
        closestToY yValue: Double,
        rounding: ChartDataSetRounding) -> ChartDataEntry?
    {
        let index = entryIndex(x: xValue, closestToY: yValue, rounding: rounding)
        if index > -1
        {
            return self[index]
        }
        return nil
    }
```

`entryIndex` returns -1, because partitioningIndex wil return `entries.count` if the predicate always returns `false`.
```
   open override func entryIndex(
        x xValue: Double,
        closestToY yValue: Double,
        rounding: ChartDataSetRounding) -> Int {
        var closest = partitioningIndex { $0.x >= xValue }
        guard closest < endIndex else { return -1 }
```

This works for the opposite scenario (xaxis.minimum < lowest data x), because the predicate will always be true, and therefore the first data entry will be returned.

A solution might be:
```
var closest = partitioningIndex { $0.x >= xValue }
closest = Swift.min(closest, entries.count - 1)
```
But this may be weird in other scenarios.

Another solution may be to update the `XBound.set` where this scenario is taken into consideration:
```
        open func set(chart: BarLineScatterCandleBubbleChartDataProvider,
                      dataSet: BarLineScatterCandleBubbleChartDataSetProtocol,
                      animator: Animator?)
        {
            let phaseX = Swift.max(0.0, Swift.min(1.0, animator?.phaseX ?? 1.0))
            
            let low = chart.lowestVisibleX
            let high = Double.minimum(chart.highestVisibleX, dataSet.xMax)
            
            let entryFrom = dataSet.entryForXValue(low, closestToY: .nan, rounding: .down)
            let entryTo = dataSet.entryForXValue(high, closestToY: .nan, rounding: .up)
            
            self.min = entryFrom == nil ? 0 : dataSet.entryIndex(entry: entryFrom!)
            self.max = entryTo == nil ? 0 : dataSet.entryIndex(entry: entryTo!)
            range = Int(Double(self.max - self.min) * phaseX)
        }
```

**Comment by iKr1s** [comment]
Someone posted a code example with a workaround in #4605.

**Comment by jeroenrietveld** [comment]
Thanks, not sure why that hasn't come up in my search..
I will close this issue. Hopefully some progress will be made on #4605.

**Comment by be-bert** [comment]
@jeroenrietveld I am having the same issue in 4.0.1.  This was not a problem in earlier builds so I think this should remain open.  Was the fix in #4605 to override `entryIndex`?

**Comment by shtoba** [comment]
@be-bert @jeroenrietveld I'm having the same issue. Has there been any update?
